tidy3d/components/grid/grid_spec.py

```diff
@@ -665,14 +787,580 @@ def _preprocessed_structures(self, structures: List[StructureType]) -> List[Stru
+            applied along that axis.
+        grid_size_in_vaccum : float
+            Grid step size in vaccum.
+
+        Returns
+        -------
+        MeshOverrideStructure
+            Unshadowed override structures for mesh refinement. If refinement doesn't need to be applied to an axis,
+            the override geometry has size=inf and dl=None along this axis.
+        """
+
+        dl = self._grid_size(grid_size_in_vacuum)
+        # override step size list
+        dl_list = [None if axis_c is None else dl for axis_c in center]
+        # override structure
+        center_geo = [0 if axis_c is None else axis_c for axis_c in center]
+        size_geo = [inf if axis_c is None else dl * self.num_cells for axis_c in center]
+        return MeshOverrideStructure(
+            geometry=Box(center=center_geo, size=size_geo), dl=dl_list, shadow=False
+        )
+
+
+class LayerRefinementSpec(Box):
+    """Specification for automatic mesh refinement and snapping in layered structures. Structure corners
+    on the cross section perpendicular to layer thickness direction can be automatically identified. Subsequently,
+    mesh is snapped and refined around the corners. Mesh can also be refined and snapped around the bounds along
+    the layer thickness direction.
+
+    Note
+    ----
+
+    Corner detection is performed on a 2D plane sitting in the middle of the layer. If the layer is finite
+    along inplane axes, corners outside the bounds are discarded.
+
+    Note
+    ----
+
+    This class only takes effect when :class:`AutoGrid` is applied.
+
+    Example
+    -------
+    >>> layer_spec = LayerRefinementSpec(axis=2, center=(0,0,0), size=(td.inf, td.inf, 1))
+
+    """
+
+    axis: Axis = pd.Field(
+        ...,
+        title="Axis",
+        description="Specifies dimension of the layer normal axis (0,1,2) -> (x,y,z).",
+    )
+
+    min_steps_along_axis: Optional[pd.PositiveFloat] = pd.Field(
+        None,
+        title="Minimal Number Of Steps Along Axis",
+        description="If not ``None`` and the thickness of the layer is nonzero, set minimal "
+        "number of steps discretizing the layer thickness.",
+    )
+
+    bounds_refinement: Optional[GridRefinement] = pd.Field(
+        None,
+        title="Mesh Refinement Factor Around Layer Bounds",
+        description="If not ``None``, refine mesh around minimum and maximum positions "
+        "of the layer along axis dimension. If `min_steps_along_axis` is also specified, "
+        "refinement here is only applied if it sets a smaller grid size.",
+    )
+
+    bounds_snapping: Optional[Literal["bounds", "lower", "upper", "center"]] = pd.Field(
+        "lower",
+        title="Placing Grid Snapping Point Along Axis",
+        description="If not ``None``, enforcing grid boundaries to pass through ``lower``, "
+        "``center``, or ``upper`` position of the layer; or both ``lower`` and ``upper`` with ``bounds``.",
+    )
```

Comment:
Same question, how does this work if the layer is too thin but ``bounds`` is selected?